Add WKApplicationManifest API

beidson opened the pull request:
#### 275b520c94280c07ae81e2769ae4f71a6c1e301b
<pre>
Add WKApplicationManifest API
<a href="https://bugs.webkit.org/show_bug.cgi?id=251298">https://bugs.webkit.org/show_bug.cgi?id=251298</a>
rdar://104630943

Reviewed by Andy Estes.

WKApplicationManifest directly reflects how the _WKApplicationManifest SPI works, but with a few differences:
1 - It is not constructible by the API client, only accessible from a WKWebView instance
2 - API-wise it is an opaque token, instead of having accessible properties
3 - It will have more private properties added to it in a followup patch that _WKApplicationManifest doesn&apos;t need

So I&apos;m creating a new object instead of just promoting the older SPI.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifest.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifest.mm: Added.
(-[WKApplicationManifest _apiObject]):
(-[WKApplicationManifest initWithApplicationManifest:]):
(nullableNSString):
(-[WKApplicationManifest _name]):
(-[WKApplicationManifest _shortName]):
(-[WKApplicationManifest _applicationDescription]):
(-[WKApplicationManifest _scope]):
(-[WKApplicationManifest _startURL]):
(-[WKApplicationManifest _themeColor]):
(-[WKApplicationManifest _displayMode]):
(-[WKApplicationManifest _icons]):
(-[WKApplicationManifest _manifestId]):
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifestInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifestPrivate.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView getApplicationManifestWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ApplicationManifest.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/259523@main">https://commits.webkit.org/259523@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e24a19324ad4b12cd374713397d4c9f5f6d68902

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105129 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14209 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38017 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114388 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109030 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5131 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97446 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110885 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26502 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7542 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/27861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7638 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47416 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6559 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9424 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rgins commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
```diff
@@ -664,6 +665,13 @@ The uniform type identifier kUTTypeWebArchive can be used get the related pasteb
 
+/*!
+@abstract Gets Web Application Manifest data from the WKWebView to be used with UIActivityViewController
+@param completionHandler A block to invoke with the available WKApplicationManfiest data.
```
```suggestion
@param completionHandler A block to invoke with the available WKApplicationManifest data.
```

rgins commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
```diff
@@ -664,6 +665,13 @@ The uniform type identifier kUTTypeWebArchive can be used get the related pasteb
+/*!
+@abstract Gets Web Application Manifest data from the WKWebView to be used with UIActivityViewController
+@param completionHandler A block to invoke with the available WKApplicationManfiest data.
+The WKApplicationManfiest argument to this block may be nil if the web page does not specify the bare minimum Web Application Manfiest data.
```
```suggestion
The WKApplicationManifest argument to this block may be nil if the web page does not specify the bare minimum Web Application Manifest data.
```

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifest.mm:
```diff
@@ -0,0 +1,118 @@
+    return *_applicationManifest;
+}
+
+-(instancetype)initWithApplicationManifest:(Ref<API::ApplicationManifest>&&)applicationManifest
```
Nit: missing a space between `-` and `(`

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifest.mm:
```diff
@@ -0,0 +1,118 @@
+#import <wtf/cocoa/VectorCocoa.h>
+
+@interface _WKApplicationManifestIcon ()
+- (instancetype)initWithCoreIcon:(const WebCore::ApplicationManifest::Icon *)icon;
```
Nit: I thought our style was to not put a space before `*` with C++ types, but maybe that's no longer true.

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifest.mm:
```diff
@@ -0,0 +1,118 @@
+    return _applicationManifest->applicationManifest().startURL;
+}
+
+- (WebCore::CocoaColor *)_themeColor
```
Nit: I thought our style was to not put a space before `*` with C++ types, but maybe that's no longer true.

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifestInternal.h:
```diff
@@ -0,0 +1,36 @@
+    RefPtr<API::ApplicationManifest> _applicationManifest;
+}
+
+-(instancetype)initWithApplicationManifest:(Ref<API::ApplicationManifest>&&)applicationManifest;
```
Nit: missing a space between `-` and `(`

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1964,6 +1965,27 @@ - (void)setMinimumViewportInset:(CocoaEdgeInsets)minimumViewportInset maximumVie
+    THROW_IF_SUSPENDED;
+#if ENABLE(APPLICATION_MANIFEST)
+    _page->getApplicationManifest([completionHandler = makeBlockPtr(completionHandler)](const std::optional<WebCore::ApplicationManifest>& manifest) {
+        if (completionHandler) {
```
Since `completionHandler` is assumed to be non-null in WKWebView.h, do we need to null-check it here? If so, should this be an early-return to reduce indentation?

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1964,6 +1965,27 @@ - (void)setMinimumViewportInset:(CocoaEdgeInsets)minimumViewportInset maximumVie
+        if (completionHandler) {
+            if (manifest) {
+                auto apiManifest = API::ApplicationManifest::create(*manifest);
+                WKApplicationManifest *wkManifest = [[WKApplicationManifest alloc] initWithApplicationManifest:WTFMove(apiManifest)];
```
Can we use `adoptNS` here to avoid the manual `-release` below?

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1964,6 +1965,27 @@ - (void)setMinimumViewportInset:(CocoaEdgeInsets)minimumViewportInset maximumVie
+                completionHandler(wkManifest);
+                [wkManifest release];
+            } else
+                completionHandler(nil);
```
WKWebView.h declares this completion handler to take a non-null WKApplicationManifest (due to `NS_ASSUME_NONNULL`). Should it be made nullable if we are going to pass a `nil` in some cases?

beidson committed:
Add WKApplicationManifest API
https://bugs.webkit.org/show_bug.cgi?id=251298
rdar://104630943

Reviewed by Andy Estes.

WKApplicationManifest directly reflects how the _WKApplicationManifest SPI works, but with a few differences:
1 - It is not constructible by the API client, only accessible from a WKWebView instance
2 - API-wise it is an opaque token, instead of having accessible properties
3 - It will have more private properties added to it in a followup patch that _WKApplicationManifest doesn't need

So I'm creating a new object instead of just promoting the older SPI.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifest.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifest.mm: Added.
(-[WKApplicationManifest _apiObject]):
(-[WKApplicationManifest initWithApplicationManifest:]):
(nullableNSString):
(-[WKApplicationManifest _name]):
(-[WKApplicationManifest _shortName]):
(-[WKApplicationManifest _applicationDescription]):
(-[WKApplicationManifest _scope]):
(-[WKApplicationManifest _startURL]):
(-[WKApplicationManifest _themeColor]):
(-[WKApplicationManifest _displayMode]):
(-[WKApplicationManifest _icons]):
(-[WKApplicationManifest _manifestId]):
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifestInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/WKApplicationManifestPrivate.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView getApplicationManifestWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorInternal.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ApplicationManifest.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/259523@main
